File: Haskell/FP-3-uzd/FP3/MExpression/Serialize.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FP3
{
    public class Serializer
    {
        public static string Serialize(List<Move> moves)
        {
            string prefix = "m[";
            string suffix = "]";
            string serializedMoves = SerializeMoves(moves);
            if(serializedMoves.Length == 0)
                return prefix + serializedMoves.Substring(0, serializedMoves.Length) + suffix;
            return prefix + serializedMoves.Substring(0, serializedMoves.Length - 1) + suffix;
        }

        private static string SerializeMoves(List<Move> moves)
        {
            return SerializeMoves1(moves, "", 0);
        }

        private static string SerializeMoves1(List<Move> moves, string acc, int i)
        {
            if (moves.Count == 0)
                return acc;
            string serializedMove = SerializeMove(moves.First(), i);
            return SerializeMoves1(moves.GetRange(1, moves.Count - 1), acc + serializedMove, i + 1);
        }

        private static string SerializeMove(Move move, int i)
        {
            string prefix = "\"" + i + "\";m[";
            string xNotation = "\"x\";";
            string yNotation = "\"y\";";
            string vNotation = "\"v\";";
            int x = move.x;
            int y = move.y;
            char v = move.z;

            return (prefix + xNotation + x + ";" + yNotation + y + ";" + vNotation + "\"" + v + "\"];" + string.Empty);
        }
    }
}

File: Haskell/FP-2-uzd/src/MExpression/Serialize.hs
module MExpression.Serialize where

import Types

serialize :: [Move] -> String
serialize moves = 
  let
    prefix = "m["
    suffix = "]"
    serializedMoves = serializeMoves moves
  in prefix ++ (take (length serializedMoves - 1) serializedMoves) ++ suffix

serializeMoves :: [Move] -> String
serializeMoves moves = serializeMoves' moves "" 0
  where
    serializeMoves' :: [Move] -> String -> Int -> String
    serializeMoves' [] acc i = acc
    serializeMoves' (h:t) acc i = serializeMoves' t (acc ++ serializedMove) (i + 1)
      where
        serializedMove = serializeMove h i

serializeMove :: Move -> Int -> String
serializeMove move i =
  let 
    prefix = "\"" ++ show i ++ "\";m["
    xNotation = "\"x\";"
    yNotation = "\"y\";"
    vNotation = "\"v\";"
    x = extractX move
    y = extractY move
    v = extractPlayer move
  in
    prefix ++ xNotation ++ show x ++ ";" ++ yNotation ++ show y ++ ";" ++ vNotation ++ "\"" ++ [v] ++ "\"];"